nemo/collections/asr/parts/mixins/ipl_mixin.py

```diff
@@ -0,0 +1,679 @@
+    """
+
+    def setup_ipl(self, model_type: str):
+        """
+        Sets up IPL configuration for the model.
+
+        Args:
+            _ipl_model_type (str): The type of model being used. Takes values "hybrid" or "ctc".
+        """
+
+        ipl_config = self.cfg.get("ipl")
+        self._ipl_params = {}
+        self._ipl_model_type = model_type
+        if ipl_config is not None:
+            if self.trainer and self.trainer.max_steps < 0:
+                raise ValueError(" For IPL to work max steps should be provided in the trainer.")
+            self._set_ipl_params(ipl_config)
+            if self.cfg.train_ds.get("is_tarred", False):
+                self._ipl_params['cache_manifest'] = []
+                self._ipl_params['all_cache_manifests'] = formulate_cache_manifest_names(
+                    self._ipl_params['manifest_filepath'], self._ipl_params['cache_prefix'], is_tarred=True
+                )
+            else:
+                if not self._ipl_params.get("cache_manifest", None):
+                    self._ipl_params['cache_manifest'] = formulate_cache_manifest_names(
+                        self._ipl_params['manifest_filepath'], self._ipl_params['cache_prefix'], is_tarred=False
+                    )
+
+    def _set_ipl_params(self, ipl_config):
+        """
+        Processes and sets IPL parameters from the configuration.
+
+        Args:
+            ipl_config (DictConfig): The configuration dictionary for IPL parameters.
+        """
+        required_params = {'m_epochs', 'manifest_filepath', 'is_tarred', 'dropout', 'n_l_epochs', 'p_cache'}
+
+        supported_params = {
+            'm_epochs',
+            'restore_pc',
+            'manifest_filepath',
+            'tarred_audio_filepaths',
+            'is_tarred',
+            'dataset_weights',
+            'dropout',
+            'limit_train_batches',
+            'cache_manifest',
+            'n_l_epochs',
+            'p_cache',
+            'cache_prefix',
+            'batch_size',
+        }
+
+        for param_name, param_value in ipl_config.items():
+            if param_name in supported_params:
+                self._ipl_params[param_name] = param_value
+            else:
+                logging.warning(f"Unsupported IPL parameter: {param_name}. This parameter will be ignored.")
+
+        # Check for missing required parameters
+        missing_params = required_params - self._ipl_params.keys()
+        if missing_params:
+            raise ValueError(f"Missing required IPL parameters: {missing_params}")
+
+        # Log a warning for any extra parameters in the configuration
+        extra_params = set(ipl_config.keys()) - supported_params
+        if extra_params:
+            logging.warning(f"Extra IPL parameters found in configuration and will be ignored: {extra_params}")
+
+    def maybe_do_ipl(self):
+        """
+        Function implements the logic of IPL algorithm.
+        """
+        if not self.cfg.get("ipl"):
```

Comment:
@titu1994 I'm not a fan of checking the cfg each pass. of the training loop. Seems off...

do we have any design qualms about just creating a setter in the mixin to flag if a model is IPL or not?